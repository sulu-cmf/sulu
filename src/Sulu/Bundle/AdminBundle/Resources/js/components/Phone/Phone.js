// @flow
import React from 'react';
import Input from '../Input';

type Props = {|
    name?: string,
    placeholder?: string,
    valid: boolean,
    value: ?string,
    onBlur?: () => void,
    onChange: (value: ?string, event: SyntheticEvent<HTMLInputElement>) => void,
|};

export default class Phone extends React.PureComponent<Props> {
    static defaultProps = {
        valid: true,
    };

    handleIconClick = () => {
        const {value} = this.props;
        if (!value) {
            return;
        }

        window.location.assign('tel:' + value);
    };

    render() {
        const {
            valid,
            name,
            placeholder,
            onBlur,
            onChange,
            value,
        } = this.props;

        return (
            <Input
                icon="su-phone"
                name={name}
                onBlur={onBlur}
                onChange={onChange}
                onIconClick={(value && value.length > 1) ? this.handleIconClick : undefined}
                placeholder={placeholder}
                type="tel"
                valid={valid}
                value={value}
            />
        );
    }
}
