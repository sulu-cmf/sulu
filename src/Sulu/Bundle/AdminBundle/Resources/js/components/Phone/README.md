The phone component can be used to get a phone number from the user.

```javascript
initialState = {value: undefined};
const onChange = (newValue) => {
    setState({value: newValue});
};

<div>
    <div style={{paddingBottom: '50px'}}>Current value: {state.value ? state.value : 'null'}</div>
    <Phone value={state.value} onChange={onChange} />
</div>
```