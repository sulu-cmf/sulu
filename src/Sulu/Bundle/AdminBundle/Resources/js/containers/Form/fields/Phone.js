// @flow
import React from 'react';
import PhoneComponent from '../../../components/Phone';
import type {FieldTypeProps} from '../../../types';

export default class Phone extends React.Component<FieldTypeProps<?string>> {
    render() {
        const {dataPath, disabled, error, onChange, onFinish, value} = this.props;

        return (
            <PhoneComponent
                disabled={!!disabled}
                id={dataPath}
                onBlur={onFinish}
                onChange={onChange}
                valid={!error}
                value={value}
            />
        );
    }
}
