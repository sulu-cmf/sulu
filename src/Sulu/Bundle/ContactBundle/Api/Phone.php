<?php

/*
 * This file is part of Sulu.
 *
 * (c) Sulu GmbH
 *
 * This source file is subject to the MIT license that is bundled
 * with this source code in the file LICENSE.
 */

namespace Sulu\Bundle\ContactBundle\Api;

use JMS\Serializer\Annotation\Groups;
use JMS\Serializer\Annotation\SerializedName;
use JMS\Serializer\Annotation\VirtualProperty;
use Sulu\Bundle\ContactBundle\Entity\Phone as PhoneEntity;
use Sulu\Bundle\ContactBundle\Entity\PhoneType as PhoneTypeEntity;
use Sulu\Component\Rest\ApiWrapper;

class Phone extends ApiWrapper
{
    public function __construct(PhoneEntity $phone, $locale)
    {
        $this->entity = $phone;
        $this->locale = $locale;
    }

    /**
     * @VirtualProperty
     * @SerializedName("id")
     * @Groups({"fullContact", "fullAccount"})
     */
    public function getId(): ?int
    {
        return $this->entity->getId();
    }

    public function setPhone(string $phone): self
    {
        $this->entity->setPhone($phone);

        return $this;
    }

    /**
     * @VirtualProperty
     * @SerializedName("phone")
     * @Groups({"fullContact", "fullAccount"})
     */
    public function getPhone(): ?string
    {
        return $this->entity->getPhone();
    }

    public function setPhoneType(PhoneTypeEntity $phoneType): self
    {
        $this->entity->setPhoneType($phoneType);

        return $this;
    }

    /**
     * @VirtualProperty
     * @SerializedName("phoneType")
     * @Groups({"fullContact", "fullAccount"})
     */
    public function getPhoneType(): ?int
    {
        return $this->entity->getPhoneType()->getId();
    }
}
