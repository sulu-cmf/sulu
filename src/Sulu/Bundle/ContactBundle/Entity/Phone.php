<?php

/*
 * This file is part of Sulu.
 *
 * (c) MASSIVE ART WebServices GmbH
 *
 * This source file is subject to the MIT license that is bundled
 * with this source code in the file LICENSE.
 */

namespace Sulu\Bundle\ContactBundle\Entity;

use JMS\Serializer\Annotation\Exclude;
use JMS\Serializer\Annotation\Groups;

/**
 * Phone.
 */
class Phone
{
    /**
     * @var string
     * @Groups({"fullAccount", "partialAccount", "fullContact", "partialContact"})
     */
    private $phone;

    /**
     * @var int
     * @Groups({"fullAccount", "partialAccount", "fullContact", "partialContact"})
     */
    private $id;

    /**
     * @var \Sulu\Bundle\ContactBundle\Entity\PhoneType
     * @Groups({"fullAccount", "fullContact"})
     */
    private $phoneType;

    /**
     * @var \Doctrine\Common\Collections\Collection
     * @Exclude
     */
    private $contacts;

    /**
     * @var \Doctrine\Common\Collections\Collection
     * @Exclude
     */
    private $accounts;

    /**
     * Constructor.
     */
    public function __construct()
    {
        $this->contacts = new \Doctrine\Common\Collections\ArrayCollection();
        $this->accounts = new \Doctrine\Common\Collections\ArrayCollection();
    }

    /**
     * Set phone.
     *
     * @param string $phone
     *
     * @return Phone
     */
    public function setPhone($phone)
    {
        $this->phone = $phone;

        return $this;
    }

    /**
     * Get phone.
     *
     * @return string
     */
    public function getPhone()
    {
        return $this->phone;
    }

    /**
     * Get id.
     *
     * @return int
     */
    public function getId()
    {
        return $this->id;
    }

    /**
     * Set phoneType.
     *
     * @param \Sulu\Bundle\ContactBundle\Entity\PhoneType $phoneType
     *
     * @return Phone
     */
    public function setPhoneType(\Sulu\Bundle\ContactBundle\Entity\PhoneType $phoneType)
    {
        $this->phoneType = $phoneType;

        return $this;
    }

    /**
     * Get phoneType.
     *
     * @return \Sulu\Bundle\ContactBundle\Entity\PhoneType
     */
    public function getPhoneType()
    {
        return $this->phoneType;
    }

    /**
     * Add contacts.
     *
     * @param \Sulu\Bundle\ContactBundle\Entity\ContactInterface $contacts
     *
     * @return Phone
     */
    public function addContact(\Sulu\Bundle\ContactBundle\Entity\ContactInterface $contacts)
    {
        $this->contacts[] = $contacts;

        return $this;
    }

    /**
     * Remove contacts.
     *
     * @param \Sulu\Bundle\ContactBundle\Entity\ContactInterface $contacts
     */
    public function removeContact(\Sulu\Bundle\ContactBundle\Entity\ContactInterface $contacts)
    {
        $this->contacts->removeElement($contacts);
    }

    /**
     * Get contacts.
     *
     * @return \Doctrine\Common\Collections\Collection
     */
    public function getContacts()
    {
        return $this->contacts;
    }

    /**
     * Add accounts.
     *
     * @param AccountInterface $account
     *
     * @return Phone
     */
    public function addAccount(AccountInterface $account)
    {
        $this->accounts[] = $account;

        return $this;
    }

    /**
     * Remove accounts.
     *
     * @param AccountInterface $account
     */
    public function removeAccount(AccountInterface $account)
    {
        $this->accounts->removeElement($account);
    }

    /**
     * Get accounts.
     *
     * @return \Doctrine\Common\Collections\Collection
     */
    public function getAccounts()
    {
        return $this->accounts;
    }
}
