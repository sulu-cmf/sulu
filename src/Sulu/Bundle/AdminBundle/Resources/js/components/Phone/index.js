// @flow
import Phone from './Phone';

export default Phone;
