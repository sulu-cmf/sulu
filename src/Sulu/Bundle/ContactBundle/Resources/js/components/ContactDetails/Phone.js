// @flow
import React from 'react';
import {Phone as PhoneComponent} from 'sulu-admin-bundle/components';
import {translate} from 'sulu-admin-bundle/utils';
import type {FormFieldTypes} from 'sulu-admin-bundle/types';
import Field from './Field';

type Props = {|
    index: number,
    onBlur: () => void,
    onPhoneChange: (index: number, phone: ?string) => void,
    onRemove: (index: number) => void,
    onTypeChange: (index: number, type: number) => void,
    phone: ?string,
    type: number,
|};

export default class Phone extends React.Component<Props> {
    static types: FormFieldTypes;

    handlePhoneChange = (phone: ?string) => {
        const {index, onPhoneChange} = this.props;

        onPhoneChange(index, phone);
    };

    render() {
        const {index, onBlur, onRemove, onTypeChange, phone, type} = this.props;

        return (
            <Field
                index={index}
                label={translate('sulu_contact.phone')}
                onRemove={onRemove}
                onTypeChange={onTypeChange}
                type={type}
                types={Phone.types}
            >
                <PhoneComponent onBlur={onBlur} onChange={this.handlePhoneChange} value={phone} />
            </Field>
        );
    }
}
