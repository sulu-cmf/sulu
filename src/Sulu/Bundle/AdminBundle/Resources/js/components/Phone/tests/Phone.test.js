// @flow
import React from 'react';
import {render, mount} from 'enzyme';
import Phone from '../Phone';

test('Phone should render', () => {
    const onChange = jest.fn();
    expect(render(<Phone onChange={onChange} value={null} />)).toMatchSnapshot();
});

test('Phone should render with placeholder', () => {
    const onChange = jest.fn();
    expect(render(<Phone onChange={onChange} placeholder="My placeholder" value={null} />)).toMatchSnapshot();
});

test('Phone should render with value', () => {
    const onChange = jest.fn();
    const value = 'test@test.com';
    expect(render(<Phone onChange={onChange} value={value} />)).toMatchSnapshot();
});

test('Phone should render null value as empty string', () => {
    const onChange = jest.fn();
    expect(render(<Phone onChange={onChange} value={null} />)).toMatchSnapshot();
});

test('Phone should render error', () => {
    const onChange = jest.fn();
    expect(render(<Phone onChange={onChange} valid={false} value={null} />)).toMatchSnapshot();
});

test('Phone should trigger callbacks correctly', () => {
    const onChange = jest.fn();
    const onBlur = jest.fn();
    const phone = mount(<Phone onBlur={onBlur} onChange={onChange} value={null} />);

    phone.find('Input').instance().props.onChange('+123', {target: {value: '+123'}});
    phone.find('Input').instance().props.onBlur();
    phone.update();
    expect(onChange).toBeCalledWith('+123', {target: {value: '+123'}});
    expect(onBlur).toBeCalled();

    expect(onBlur).toHaveBeenCalledTimes(1);
    expect(onChange).toHaveBeenCalledTimes(1);
});

test('Phone should not set onIconClick when value is not set', () => {
    const onChange = jest.fn();
    const onBlur = jest.fn();
    const phone = mount(<Phone onBlur={onBlur} onChange={onChange} value={null} />);

    expect(phone.find('Input').prop('onIconClick')).toBeUndefined();
});

test('Phone should set onIconClick when value is set', () => {
    const onChange = jest.fn();
    const onBlur = jest.fn();
    const phone = mount(<Phone onBlur={onBlur} onChange={onChange} value={'+123'} />);

    expect(phone.find('Input').prop('onIconClick')).toBeInstanceOf(Function);
});

test('Phone should set onIconClick when value is valid and window should be opened', () => {
    window.location.assign = jest.fn();

    const onChange = jest.fn();
    const onBlur = jest.fn();
    const email = mount(<Phone onBlur={onBlur} onChange={onChange} value={'+123'} />);

    const onIconClickFunction = email.find('Input').prop('onIconClick');
    expect(onIconClickFunction).toBeInstanceOf(Function);
    onIconClickFunction.call();
    expect(window.location.assign).toBeCalledWith('tel:+123');
});
