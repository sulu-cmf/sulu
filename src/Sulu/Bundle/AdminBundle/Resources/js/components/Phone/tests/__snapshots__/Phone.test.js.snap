// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Phone should render 1`] = `
<label
  class="input default left"
>
  <div
    class="prependedContainer default"
  >
    <span
      aria-label="su-phone"
      class="icon default su-phone"
    />
  </div>
  <input
    type="tel"
    value=""
  />
</label>
`;

exports[`Phone should render error 1`] = `
<label
  class="input default left error"
>
  <div
    class="prependedContainer default"
  >
    <span
      aria-label="su-phone"
      class="icon default su-phone"
    />
  </div>
  <input
    type="tel"
    value=""
  />
</label>
`;

exports[`Phone should render null value as empty string 1`] = `
<label
  class="input default left"
>
  <div
    class="prependedContainer default"
  >
    <span
      aria-label="su-phone"
      class="icon default su-phone"
    />
  </div>
  <input
    type="tel"
    value=""
  />
</label>
`;

exports[`Phone should render with placeholder 1`] = `
<label
  class="input default left"
>
  <div
    class="prependedContainer default"
  >
    <span
      aria-label="su-phone"
      class="icon default su-phone"
    />
  </div>
  <input
    placeholder="My placeholder"
    type="tel"
    value=""
  />
</label>
`;

exports[`Phone should render with value 1`] = `
<label
  class="input default left"
>
  <div
    class="prependedContainer default"
  >
    <span
      aria-label="su-phone"
      class="icon default iconClickable su-phone clickable"
      role="button"
      tabindex="0"
    />
  </div>
  <input
    type="tel"
    value="test@test.com"
  />
</label>
`;
